components/core/src/clp_s/JsonFileIterator.cpp

```diff
@@ -78,52 +88,51 @@ bool JsonFileIterator::get_json(simdjson::ondemand::document_stream::iterator& i
         m_first_read = false;
     }
 
-    size_t patch_truncated_bytes = 0;
     if (m_doc_it != m_stream.end()) {
-        if (m_doc_it.error() == simdjson::error_code::SUCCESS) {
+        if (simdjson::error_code::SUCCESS == m_doc_it.error()) {
             it = m_doc_it;
+
+            // there is a bug in simdjson where when invalid UTF8 is encountered at the end of the
+            // stream truncated bytes isn't set correctly. To work around this limitation we keep
+            // track of the start of the next document. The bytes after the last valid document are
+            // naturally the truncated bytes for the stream.
+            m_next_document_position = m_doc_it.current_index() + m_doc_it.source().size();
             return true;
         } else if (m_doc_it.error() == simdjson::error_code::UTF8_ERROR) {
-            patch_truncated_bytes
-                    = reverse_search_newline_truncated_bytes(m_doc_it.current_index());
+            m_truncated_bytes = m_buf_occupied - m_next_document_position;
+        } else {
+            m_error_code = m_doc_it.error();
+            return false;
         }
     } else if (m_eof) {
+        // if we iterate to the end of the stream without error after hitting eof it should be safe
+        // to ask the stream iterator for number of truncated bytes
+        m_truncated_bytes = m_stream.truncated_bytes();
         return false;
+    } else {
+        m_truncated_bytes = m_stream.truncated_bytes();
     }
 
-    // there is a bug in simdjson where when invalid utf8 is encountered
-    // at the end of the stream truncated bytes isn't set correctly.
-    // Work around this limitation by manually searching for the start
-    // of the erroring document and stetting truncated bytes appropriately
-    if (patch_truncated_bytes == 0) {
-        read_new_json(m_stream.truncated_bytes());
-    } else {
-        read_new_json(patch_truncated_bytes);
+    if (m_truncated_bytes > 0) {
+        ++m_doc_it;
+        // UTF8 error wasn't at the end of the stream -- terminate
+        if (m_doc_it != m_stream.end()) {
+            m_error_code = simdjson::error_code::UTF8_ERROR;
```

Comment:
Just curious how can the code enters this branch

Reply:
The intention was for this code to execute once we've hit a UTF8 error but we don't know yet whether it is in the middle of the stream or at the end of the stream (and possibly just a truncated document). I added the second way of setting truncated bytes after I wrote this code block to fix a bug, and am realizing that this flow is pretty confusing now.

I think I'll rewrite this to be clearer/more explicit about what its trying to do.

Reply:
Oh yeah, that's right.